src/one/nio/http/HttpSession.java

```diff
@@ -102,35 +110,112 @@ protected void handleSocketClosed() {
+        parsing.setBody(body);
+
+        // Start consuming the body
+        requestBodyOffset = Math.min(bufferLength - bufferOffset, body.length);
+        System.arraycopy(
+                buffer,
+                bufferOffset,
+                body,
+                0,
+                requestBodyOffset);
+
+        return requestBodyOffset;
+    }
+
+    protected void handleParsedRequest() throws IOException {
+        if (handling == null) {
+            server.handleRequest(handling = parsing, this);
+        } else if (pipeline.size() < MAX_PIPELINE_LENGTH) {
+            pipeline.addLast(parsing);
+        } else {
+            throw new IOException("Pipeline length exceeded");
+        }
+        parsing = null;
+        requestBodyOffset = 0;
+    }
+
     protected int processHttpBuffer(byte[] buffer, int length) throws IOException, HttpException {
-        int lineStart = 0;
-        for (int i = 0; i < length; i++) {
+        int i = 0; // Current position in the buffer
+
+        if (parsing != null && parsing.getBody() != null) { // Resume consuming request body
+            final byte[] body = parsing.getBody();
+            i = Math.min(length, body.length - requestBodyOffset);
+            System.arraycopy(buffer, 0, body, requestBodyOffset, i);
+            requestBodyOffset += i;
+            if (requestBodyOffset < body.length) {
+                // All the buffer copied to body, but that is not enough -- wait for next data
+                return length;
+            } else {
+                // Process current request
+                if (closing) {
+                    return i;
+                } else {
+                    handleParsedRequest();
+                }
+            }
+        }
+
+        int lineStart = i;
+        for (; i < length; i++) {
             if (buffer[i] != '\n') continue;
 
             int lineLength = i - lineStart;
             if (i > 0 && buffer[i - 1] == '\r') lineLength--;
 
+            // Skip '\n'
+            i++;
+
             if (parsing == null) {
                 parsing = parseRequest(buffer, lineStart, lineLength);
             } else if (lineLength > 0) {
                 if (parsing.getHeaderCount() < MAX_HEADERS) {
                     parsing.addHeader(Utf8.read(buffer, lineStart, lineLength));
                 }
-            } else {
+            } else { // Empty line -- there is next request or body of the current request
+                final String contentLengthValue = parsing.getHeader("Content-Length: ");
+                if (contentLengthValue != null) { // Start parsing request body
+                    final int contentLength = Integer.valueOf(contentLengthValue);
+                    i += startParsingRequestBody(contentLength, buffer, i, length);
+                    if (requestBodyOffset < parsing.getBody().length) {
+                        // Consumed all the buffer data, but some bytes are still left
```

Comment:
"bytes are still left" sounds unclear. Where are they left?
You mean, the body is not yet completly read, right?